File: java/year2021/month5/no300/LongestIncreasingSubsequence.java
package year2021.month5.no300;

public class LongestIncreasingSubsequence {
    public static void main(String[] args) {
        int[] nums1 = {10, 9, 2, 5, 3, 7, 101, 18};
        int[] nums2 = {0, 1, 0, 3, 2, 3};
        int[] nums3 = {7, 7, 7, 7, 7, 7, 7};
        System.out.println(lengthOfLIS1(nums1));
        System.out.println(lengthOfLIS1(nums2));
        System.out.println(lengthOfLIS1(nums3));
    }

    private static int lengthOfLIS1(int[] nums) {
        /*贪心 + 二分
         * 要使得严格递增子序列尽可能长，那么我们需要子序列尽可能增长得缓慢
         * 用一个数组来记录当前最长且末位最小的严格递增子序列
         * 若新元素大于数组末尾元素，则将新元素 append 到数组末尾
         * 若新元素小于等于数组末尾元素，
         * 则用二分法找到 数组中最小且大于等于新元素的元素下标
         * 将其替换为新元素
         * （可以覆盖到原最长子序列的元素，若替换后的元素属于新的最长子序列
         * 那么替换元素的下标后的元素也必会被替换）
         * 结果为数组长度
         * */
        // lis 单调递增（不减）
        int[] lis = new int[nums.length];
        lis[0] = nums[0];
        int size = 1;
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] > lis[size - 1]) {
                lis[size] = nums[i];
                size++;
                continue;
            }
            // 当前元素不大于 lis 数组末尾元素
            int left = 0;
            int right = size - 1;
            int target = nums[i];
            while (left < right) {
                int mid = left + (right - left) / 2;
                if (lis[mid] >= target) {
                    right = mid;
                } else {
                    left = mid + 1;
                }
            }
            lis[left] = target;
        }
        return size;
    }

    public static int lengthOfLIS(int[] nums) {
        /*DP
         * dp[i]为从[0,i]数组中包含下标 i 的元素的最长严格递增子序列的长度
         * dp[i] = max{dp[j]} + 1, j < i && nums[j] < nums[i]
         * */
        int[] dp = new int[nums.length];
        int ans = 1;
        dp[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            dp[i] = 1;
            for (int j = 0; j < i; j++) {
                if (nums[j] >= nums[i]) {
                    continue;
                }
                dp[i] = Math.max(dp[i], dp[j] + 1);
            }
            ans = Math.max(ans, dp[i]);
        }
        return ans;
    }
}

/*
给你一个整数数组 nums ，找到其中最长严格递增子序列的长度。

子序列是由数组派生而来的序列，删除（或不删除）数组中的元素而不改变其余元素的顺序。例如，[3,6,2,7] 是数组 [0,3,1,6,2,2,7] 的子序列。

 
示例 1：

输入：nums = [10,9,2,5,3,7,101,18]
输出：4
解释：最长递增子序列是 [2,3,7,101]，因此长度为 4 。
示例 2：

输入：nums = [0,1,0,3,2,3]
输出：4
示例 3：

输入：nums = [7,7,7,7,7,7,7]
输出：1
 

提示：

1 <= nums.length <= 2500
-104 <= nums[i] <= 104
 

进阶：

你可以设计时间复杂度为 O(n2) 的解决方案吗？
你能将算法的时间复杂度降低到 O(n log(n)) 吗?

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/longest-increasing-subsequence
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
*/

File: go/year2025/month4/no300/LongestIncreasingSubsequence.go
package main

func main() {
	nums1 := []int{10, 9, 2, 5, 3, 7, 101, 18}
	nums2 := []int{0, 1, 0, 3, 2, 3}
	nums3 := []int{7, 7, 7, 7, 7, 7, 7}
	println(lengthOfLIS(nums1)) // 4
	println(lengthOfLIS(nums2)) // 4
	println(lengthOfLIS(nums3)) // 1
}

/*
单调队列，times: O(nlogn), space: O(n)
思路：用一个严格单调递增的队列来维护当前的最长递增子序列，同时让子序列上升的尽可能慢，具体如下
  - 如果当前元素 > 队列的最后一个元素，则直接在对尾追加入队列
  - 如果当前元素 <= 队列的最后一个元素，则需要找到队列中第一个大于等于当前元素的元素，替换掉它，并且从该元素位置开始继续维护单调递增队列

对于找到队列中第一个大于等于当前元素的元素，可以使用二分查找
结果：队列长度
Refer Link: https://writings.sh/post/longest-increasing-subsequence-revisited
*/
func lengthOfLIS(nums []int) int {
	queue := make([]int, 0)
	for _, num := range nums {
		queueLen := len(queue)
		if queueLen > 0 && num <= queue[queueLen-1] {
			// 如果当前元素 <= 队列的最后一个元素，则需要找到队列中第一个大于等于当前元素的元素，替换掉它
			left, right := 0, queueLen-1
			for left < right {
				mid := left + (right-left)/2 // 向 left 靠拢
				if queue[mid] >= num {
					right = mid
				} else {
					//queue[mid] < num
					left = mid + 1
				}
			}
			queue[left] = num
			continue
		}
		queue = append(queue, num)
	}
	return len(queue)
}

/*
DP, tims: O(n^2), space: O(n)
状态定义：dp[i] 为以 nums[i] 结尾的最长递增子序列的长度
状态转移方程：dp[i] = max(dp[j] + 1), (j < i && nums[j] < nums[i])
状态初始化：dp[i] = 1
*/
func lengthOfLIS1(nums []int) int {
	dp := make([]int, len(nums))
	ans := 0
	for i := 0; i < len(nums); i++ {
		dp[i] = 1
		for j := 0; j < i; j++ {
			if nums[j] < nums[i] {
				dp[i] = max(dp[i], dp[j]+1)
			}
		}
		ans = max(ans, dp[i])
	}
	return ans
}

/*
给你一个整数数组 nums ，找到其中最长严格递增子序列的长度。
子序列 是由数组派生而来的序列，删除（或不删除）数组中的元素而不改变其余元素的顺序。例如，[3,6,2,7] 是数组 [0,3,1,6,2,2,7] 的子序列。

示例 1：
输入：nums = [10,9,2,5,3,7,101,18]
输出：4
解释：最长递增子序列是 [2,3,7,101]，因此长度为 4 。

示例 2：
输入：nums = [0,1,0,3,2,3]
输出：4

示例 3：
输入：nums = [7,7,7,7,7,7,7]
输出：1

提示：
1 <= nums.length <= 2500
-104 <= nums[i] <= 104

进阶：
你能将算法的时间复杂度降低到 O(n log(n)) 吗?
*/
